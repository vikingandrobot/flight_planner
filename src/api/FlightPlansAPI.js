/**
 * Mock api functions to simulate a call to the BE and saving in DB.
 */

import { v4 as uuidv4 } from 'uuid';

const DEFAULT_PLANS = [
  {
    id: uuidv4(),
    title: 'Rolex Learning Center',
    points: [
      [731340.3154337693, 5863703.367232675],
      [731338.9386689285, 5863913.189770418],
      [731041.1223817954, 5863917.498865569],
      [731037.8860904165, 5863695.899326418],
      [731181.617955783, 5863692.388874074],
      [731184.6992866171, 5863816.190429365],
    ],
  },
  {
    id: uuidv4(),
    title: 'CathÃ©drale de Lausanne',
    points: [
      [738723.6704740375, 5864513.580420917],
      [738663.343385889, 5864522.235484955],
      [738650.7084765892, 5864448.536886312],
      [738598.5774509882, 5864455.650035162],
      [738604.9759867397, 5864530.898794529],
      [738560.8651189235, 5864542.137459774],
    ],
  }
];

let inMemoryDB = [];
inMemoryDB = inMemoryDB.concat(DEFAULT_PLANS);

const WAIT_TIME = 1000;

export function get() {
  return new Promise((resolve) => {
    setTimeout(() => { resolve(inMemoryDB); }, WAIT_TIME);
  })
}

export function post(flightPlan) {
  inMemoryDB = inMemoryDB.concat({ ...flightPlan, id: uuidv4() });

  return new Promise((resolve) => {
    setTimeout(() => { resolve(); }, WAIT_TIME);
  })
}
