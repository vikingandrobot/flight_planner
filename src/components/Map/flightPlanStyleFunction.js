import { Stroke, Style, Fill } from 'ol/style';
import { Circle } from 'ol/geom';

// Received the vector feature and returns the corresponding styles.
export default function flightPlanStyleFunction(feature) {
  const styles = [
    // linestring
    new Style({
      stroke: new Stroke({
        color: '#31a4eb',
        width: 3,
      }),
      zIndex: 0,
    }),
  ];

  const coordinates = feature.getGeometry().getCoordinates();

  return styles.concat(coordinates.map(coordinate => new Style({
    geometry: new Circle(coordinate, 4),
    fill: new Fill({
      color: '#6bc7ff',
    }),
    stroke: new Stroke({
      color: '#31a4eb',
      width: 2,
    }),
    zIndex: 1,
  })));
}
