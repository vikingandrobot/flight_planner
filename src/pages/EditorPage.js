import React from 'react';

import styles from './EditorPage.module.scss';

function EditorPage() {
  return (
    <div className={styles.editorPage}>
      <h1>Flight Planner</h1>
    </div>
  );
}

export default EditorPage;
