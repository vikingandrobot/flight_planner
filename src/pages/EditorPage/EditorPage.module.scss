@import '../../styles/colors.scss';

.editorPage {
  background-color: $veryLightBackground;
  width: 100vw;
  height: 100vh;
  color: $darkTextColor;
  display: flex;
  flex-direction: column;

  .pageHeader {
    padding: 5px 10px;
    background: $darkBlueBackground;
    height: auto;
    h1 {
      color: $veryLightTextColor;
      margin: 0;
    }
  }

  .editorContent {
    position: relative;
    flex-grow: 1;
    display: flex;
    flex-direction: column;

    .editorHeader {
      display: flex;
      position: relative;
      padding: 5px 10px;
      background-color: $veryLightBackground;

      & > div {
        flex-grow: 1;
      }
    }

    .actionsContainer {
      display: flex;
      flex-direction: row-reverse;
    }

    .editorMap {
      flex-grow: 1;
      width: 100%;
      height: 100%;
    }
    .editorFlightList {
      height: 100%;
    }
  }
}
