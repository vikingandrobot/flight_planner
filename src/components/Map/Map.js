import React, { useRef, useEffect, useCallback } from 'react';
import PropTypes from 'prop-types';
import 'ol/ol.css';
import OlMap from 'ol/Map';
import View from 'ol/View';
import TileLayer from 'ol/layer/Tile';
import VectorLayer from 'ol/layer/Vector';
import Feature from 'ol/Feature';
import LineString from 'ol/geom/LineString';
import OSM from 'ol/source/OSM';
import VectorSource from 'ol/source/Vector';
import ResizeObserver from 'resize-observer-polyfill';

import flightPlanStyleFunction from './flightPlanStyleFunction';

import styles from './Map.module.scss';

function Map({
  className = '',
  center = [0, 0],
  zoom = 1,
  onClick,
  fitViewToFlightPlan,
  flightPlan = null,
}) {
  const targetRef = useRef(null);
  const mapRef = useRef(null);
  const viewRef = useRef(null);
  const flightPlanSourceRef = useRef(new VectorSource());
  const sizeObserverRef = useRef();

  const hasFlightPlan = !!flightPlan;

  const handleClick = useCallback((e) => {
    if (onClick) {
      onClick(e.coordinate);
    }
  }, [onClick]);

  useEffect(() => {
    function createMap() {
      if (targetRef.current && !mapRef.current) {
        viewRef.current = new View({ center, zoom });

        mapRef.current = new OlMap({
          view: viewRef.current,
          layers: [
            new TileLayer({
              source: new OSM(),
            }),
            new VectorLayer({
              source: flightPlanSourceRef.current,
              style: flightPlanStyleFunction,
            })
          ],
          target: targetRef.current,
        });

        // Size observer for the map
        sizeObserverRef.current = new ResizeObserver(() => {
          mapRef.current.updateSize()
        });
        sizeObserverRef.current.observe(targetRef.current);
      }
    }

    createMap();

    return () => {
      // Clean up
      if (sizeObserverRef.current) {
        sizeObserverRef.current.disconnect();
      }
    }
  }, []);

  useEffect(() => {
    if (mapRef.current && onClick) {
      mapRef.current.un('click', handleClick);
      mapRef.current.on('click', handleClick);
    }

    return () => {
      mapRef.current.un('click', handleClick);
    }
  }, [onClick, handleClick]);

  useEffect(() => {
    flightPlanSourceRef.current.clear();
    if (flightPlan) {
      const feature = new Feature({
        geometry: new LineString(flightPlan.points),
      });
      flightPlanSourceRef.current.addFeature(feature);

      if (fitViewToFlightPlan && viewRef.current) {
        viewRef.current.fit(feature.getGeometry(), {padding: [100, 100, 100, 100]});
      }
    }
  }, [hasFlightPlan, JSON.stringify(flightPlan?.points)])

  return (
    <div ref={targetRef} className={`${styles.mapContainer} ${className}`} />
  )
}

Map.propTypes = {
  // A class name to pass to the map container to modify its style
  className: PropTypes.string,
  // Center of the map
  center: PropTypes.arrayOf(PropTypes.number),
  zoom: PropTypes.number,
  fitViewToFlightPlan: PropTypes.bool,
  flightPlan: PropTypes.shape({
    points: PropTypes.arrayOf(PropTypes.arrayOf(PropTypes.number)),
  })
};

export default Map;
