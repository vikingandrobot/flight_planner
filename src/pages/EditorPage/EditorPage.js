import React, { useState, useEffect } from 'react';
import { Button, Input, Typography } from 'antd';
import { PlusCircleOutlined, CheckCircleOutlined } from '@ant-design/icons';
import { fromLonLat } from 'ol/proj';

import { get, post } from '../../api/FlightPlansAPI';
import Map from '../../components/Map/Map';
import FlightPlanList from '../../components/FlightPlanList/FlightPlanList';

import styles from './EditorPage.module.scss';

const { Title } = Typography;

const LONGITUDE = 6.5685000;
const LATITUDE = 46.5185000;

const center = fromLonLat([LONGITUDE, LATITUDE]);
const zoom = 17;

function NewFlightActions({ onCreate, onCancel, disabled }) {
  return (
    <>
      <Button
        type="primary"
        onClick={onCreate}
        disabled={disabled}
        icon={<CheckCircleOutlined />}
      >
        Create Flight Plan
      </Button>
      <Button
        onClick={onCancel}
      >
        Cancel
      </Button>
    </>
  )
}

function DefaultActions({ onNew }) {
  return (
    <Button
      type="primary"
      icon={<PlusCircleOutlined />}
      onClick={onNew}
    >
      New Flight Plan
    </Button>
  )
}

function EditorPage() {
  const [isLoadingData, setIsLoadingData] = useState(false);
  const [isCreating, setIsCreating] = useState(false);
  const [flightPlans, setFlightPlans] = useState([]);
  const [selectedFlightPlanId, setSelectedFlightPlanId] = useState(null);
  const [draftFlightPlan, setDraftFlightPlan] = useState(null);

  useEffect(() => {
    setIsLoadingData(true);
    get()
      .then(setFlightPlans)
      .finally(() => setIsLoadingData(false));
  }, []);

  function handleNew() {
    setIsCreating(true);
    setSelectedFlightPlanId(null);
    setDraftFlightPlan({ title: null, points: [] });
  }

  function handleCreate() {
    setIsLoadingData(true);
    post(draftFlightPlan)
      .then(get)
      .then((plans) => {
        setDraftFlightPlan(null);
        setIsCreating(false);
        setFlightPlans(plans);

      })
      .finally(() => {
        setIsLoadingData(false);
      });
  }

  function handleCancel() {
    setIsCreating(false);
    setDraftFlightPlan(null);
  }

  function handleSelect({ id }) {
    setSelectedFlightPlanId(id);
  }

  function handleTitleChange(e) {
    setDraftFlightPlan(prevDraft => ({ ...prevDraft, title: e.target.value }));
  }

  function handleDraftClick(coordinate) {
    setDraftFlightPlan(prevDraft => ({ ...prevDraft, points: [...draftFlightPlan.points, coordinate] }));
  }

  return (
    <div className={styles.editorPage}>
      <header className={styles.pageHeader}>
        <Title level={1}>Flight Planner</Title>
      </header>
      <section className={styles.editorContent}>
        <header className={styles.editorHeader}>
          <div>
            {isCreating && (
              <Input
                placeholder="Flight Plan name"
                onChange={handleTitleChange}
              />
            )}
          </div>

          <div className={styles.actionsContainer}>
            {isCreating ?
            <NewFlightActions
              onCreate={handleCreate}
              onCancel={handleCancel}
              disabled={!draftFlightPlan || draftFlightPlan.points.length < 2}
            /> :
            <DefaultActions onNew={handleNew} />}
          </div>
        </header>
        <div className={styles.editorMapContainer}>
          <Map
            center={center}
            className={styles.editorMap}
            fitViewToFlightPlan={!isCreating}
            flightPlan={isCreating ? draftFlightPlan : (selectedFlightPlanId && flightPlans.find(fp => fp.id === selectedFlightPlanId))}
            onClick={isCreating ? handleDraftClick : null}
            zoom={zoom}
          />
          <FlightPlanList
            className={styles.editorFlightList}
            flightPlans={flightPlans}
            loading={isLoadingData}
            onRowClick={handleSelect}
          />
        </div>
      </section>
    </div>
  );
}

export default EditorPage;
