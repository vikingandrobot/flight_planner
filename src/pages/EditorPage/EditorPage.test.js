import { render, screen, act } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import matchMediaPolyfill from '../../../testUtils/matchMediaPolyfill';
import * as FlightPlansAPIImport from '../../api/FlightPlansAPI';
import * as MapImport from '../../components/Map/Map';
import EditorPage from './EditorPage';


/*
 * Since most of the Map interactions happen outside of React and in scoped
 * environment, we're gonna mock the Map and display the props instead.
 */
jest.mock('../../components/Map/Map', () => ({
  __esModule: true,
  default: function MapMock(props) {
    const { flightPlan } = props;
    return (
      <>
        <div>MapMock</div>
        <div>{flightPlan?.id}</div>
      </>
    );
  },
}));

const DEFAULT_PLANS = [
  {
    id: 'abcd-1234',
    title: 'Rolex Learning Center',
    points: [
      [731340.3154337693, 5863703.367232675],
      [731338.9386689285, 5863913.189770418],
    ],
  },
  {
    id: 'efgh-5678',
    title: 'Cathédrale de Lausanne',
    points: [
      [738723.6704740375, 5864513.580420917],
      [738663.343385889, 5864522.235484955],
      [738650.7084765892, 5864448.536886312],
    ],
  }
];

async function renderEditorPageWithChildrenSpies() {
  const constructorSpy = jest.spyOn(MapImport, 'default');

  let result;
  await act(async () => {
    result = render(<EditorPage />);
  });


  return {
    ...result,
    mapLatestProps: () => constructorSpy.mock.calls[constructorSpy.mock.calls.length - 1][0],
  };
}

describe('EditorPage', () => {

  let plans = DEFAULT_PLANS.slice();

  beforeEach(() => {
    matchMediaPolyfill();
    plans = DEFAULT_PLANS.slice();
    jest.spyOn(FlightPlansAPIImport, 'get').mockResolvedValue(plans);
    jest.spyOn(FlightPlansAPIImport, 'post').mockImplementation((fp) => {
      plans.push(fp);
      return Promise.resolve();
    });
  });

  test('should render the editor page', async () => {
    let result;
    await act(async () => {
      result = render(<EditorPage />);
    });
    expect(result.container).toMatchSnapshot();
  })

  describe('Existing Flight Plans', () => {
    test('should load and render the list of flight plans', async () => {
      render(<EditorPage />);

      const firstFlightPlan = await screen.findByText('Rolex Learning Center');
      expect(firstFlightPlan).toBeInTheDocument();
      expect(screen.getByText('Cathédrale de Lausanne')).toBeInTheDocument();
    });

    test('should display the selected flight plan in the map', async () => {
      render(<EditorPage />);

      await screen.findByText('Rolex Learning Center');
      userEvent.click(screen.getAllByText('View')[1]);

      expect(screen.getByText('efgh-5678')).toBeInTheDocument();
      userEvent.click(screen.getAllByText('View')[0]);

      expect(screen.getByText('abcd-1234')).toBeInTheDocument();
    });
  });

  describe('Creating a Flight Plan', () => {
    test('should display a button to draw a new Flight Plan', async () => {
      await act(async () => {
        render(<EditorPage />);
      });
      expect(screen.getByText('New Flight Plan')).toBeInTheDocument();
    });

    test('should display a button to cancel and a new button to create the new Flight Plan', async () => {
      await act(async () => {
        render(<EditorPage />);
      });

      expect(screen.queryByText('Cancel')).not.toBeInTheDocument();
      expect(screen.queryByText('Create Flight Plan')).not.toBeInTheDocument();

      userEvent.click(screen.getByText('New Flight Plan'));
      expect(screen.getByText('Cancel')).toBeInTheDocument();
      expect(screen.getByText('Create Flight Plan')).toBeInTheDocument();
    });

    test('should disable the create button if there are less that 2 points in the flight plans', async () => {
      const postSpy = jest.spyOn(FlightPlansAPIImport, 'post');

      await act(async () => {
        render(<EditorPage />);
      });
      userEvent.click(screen.getByText('New Flight Plan'));
      userEvent.click(screen.getByText('Create Flight Plan'));

      expect(postSpy).not.toHaveBeenCalled();
      expect(screen.queryByText('Please enter a name for your Flight Plan')).not.toBeInTheDocument();
    });

    test('should create a new flight plan', async () => {
      const postSpy = jest.spyOn(FlightPlansAPIImport, 'post');
      const { mapLatestProps } = await renderEditorPageWithChildrenSpies();

      userEvent.click(screen.getByText('New Flight Plan'));

      // Add some points
      act(() => {
        mapLatestProps().onClick([111, 222]);
      });

      act(() => {
        mapLatestProps().onClick([333, 444]);
      });
      // Type a title
      userEvent.type(screen.getByRole('textbox'), 'My new title');

      userEvent.click(screen.getByText('Create Flight Plan'));

      expect(postSpy).toHaveBeenCalledWith({
        title: 'My new title',
        points: [
          [111, 222],
          [333, 444],
        ],
      });
      const newlyCreated = await screen.findByText('My new title');
      expect(newlyCreated).toBeInTheDocument();
    });
  });
});
