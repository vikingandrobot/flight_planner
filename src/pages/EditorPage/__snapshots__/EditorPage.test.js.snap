// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditorPage should render the editor page 1`] = `
<div
  className="editorPage"
>
  <header
    className="pageHeader"
  >
    <Title
      level={1}
    >
      Flight Planner
    </Title>
  </header>
  <section
    className="editorContent"
  >
    <header
      className="editorHeader"
    >
      <div />
      <div
        className="actionsContainer"
      >
        <DefaultActions
          onNew={[Function]}
        />
      </div>
    </header>
    <div
      className="editorMapContainer"
    >
      <Map
        center={
          Array [
            731202.0752756175,
            5863832.126542759,
          ]
        }
        className="editorMap"
        fitViewToFlightPlan={true}
        flightPlan={null}
        onClick={null}
        zoom={17}
      />
      <FlightPlanList
        className="editorFlightList"
        flightPlans={Array []}
        loading={false}
        onRowClick={[Function]}
      />
    </div>
  </section>
</div>
`;
